### PR TITLE
Add new metric usages and update RunningAverage accordingly

### PR DESCRIPTION
**Pull request by sadra-barikbin**
Description:

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

### CODE REVIEW

File: ignite/metrics/metric.py [hunk 1]
```diff
@@ -346,10 +435,16 @@ def _check_usage(self, usage: Union[str, MetricUsage]) -> MetricUsage:
         if isinstance(usage, str):
             if usage == EpochWise.usage_name:
                 usage = EpochWise()
+            elif usage == RunningEpochWise.usage_name:
```
Comment (vfdev-5) [inline ignite/metrics/metric.py:438]: Let's make this code more generic, let's register all usages in a private list and here we can just iterate over this list.

File: ignite/metrics/running_average.py [hunk 2]
```diff
@@ -139,16 +142,62 @@ def compute(self) -> Union[torch.Tensor, float]:
...
+
+        ``RunningAverage`` retrieves ``engine.state.output`` at ``usage.ITERATION_COMPLETED`` if the ``src`` is not
+        given and it's computed and updated using ``src`, by manually calling its ``compute`` method, or
+        ``engine.state.output`` at ``usage.COMPLETED``event.
+        Also if ``src`` is given, it is updated at ``usage.ITERATION_COMPLETED``, but its reset event is determined by
+        ``usage`` type. If ``isinstance(usage, BatchWise)`` holds true, ``src`` is reset on ``BatchWise().STARTED``,
+        otherwise on ``EpochWise().STARTED`` if ``isinstance(usage, EpochWise)``.
+        """
+        usage = self._check_usage(usage)
+        has_src = hasattr(self, "src") and isinstance(self.src, Metric)
```
Comment (vfdev-5) [inline ignite/metrics/running_average.py:186]: Let's set `self.src = None` to indicate that source is Metric instead of checking attribute existence

File: ignite/metrics/running_average.py [hunk 3]
```diff
@@ -139,16 +142,62 @@ def compute(self) -> Union[torch.Tensor, float]:
...
+
+        src_usage = EpochWise() if isinstance(usage, EpochWise) else BatchWise()
+        if has_src and not engine.has_event_handler(self.src.started, src_usage.STARTED):
+            engine.add_event_handler(src_usage.STARTED, self.src.started)
+            engine.add_event_handler(usage.ITERATION_COMPLETED, self.src.iteration_completed)
```
Comment (vfdev-5) [inline ignite/metrics/running_average.py:191]: I'm not sure to understand why we should have `src_usage` and `usage` ?

**Review by vfdev-5 — COMMENTED** [review]
Thanks a lot for the PR, @sadra-barikbin !
I left several comments to improve the code and the API

File: ignite/metrics/running_average.py [hunk 4]
```diff
@@ -139,16 +142,62 @@ def compute(self) -> Union[torch.Tensor, float]:
...
-        if self.epoch_bound:
-            # restart average every epoch
-            engine.add_event_handler(Events.EPOCH_STARTED, self.started)
-        else:
-            engine.add_event_handler(Events.STARTED, self.started)
-        # compute metric
-        engine.add_event_handler(Events.ITERATION_COMPLETED, self.iteration_completed)
-        # apply running average
-        engine.add_event_handler(Events.ITERATION_COMPLETED, self.completed, name)
+    def attach(self, engine: Engine, name: str, usage: Union[str, MetricUsage] = SingleEpochRunningBatchWise()) -> None:
```
[the remark this answers is not in the document]
Comment (sadra-barikbin) [reply ignite/metrics/running_average.py:145]: I set it to be like what was the default earlier i.e. `epoch_bound=True`

File: ignite/metrics/running_average.py (hunk 2 above)
Comment (sadra-barikbin) [reply ignite/metrics/running_average.py:186]: We can't simply use `if self.src` in `attach` since `self.src` gets value in `_output_update` when there's no metric. This makes problem in the case of calling `attach` again.

By the way, why `detach` has not been implemented for `RunningAverage`?

File: ignite/metrics/running_average.py (hunk 3 above)
Comment (sadra-barikbin) [reply ignite/metrics/running_average.py:191]: I note that since `RunningAverage` gets both computed and updated at the same event, I only kept its compute event in the new usages and set the `ITERATION_COMPLETED` attribute of them to be the update event of the `src` metric. As among these three usages, `RunningEpochWise`s own update event (`EPOCH_COMPLETED`) is not the same as that of the `src` metric (`ITERATION_COMPLETED`). Otherwise I should have done what I did with reset event, with update event as well.

File: ignite/metrics/running_average.py (hunk 2 above)
Comment (vfdev-5) [reply ignite/metrics/running_average.py:186]: > We can't simply use if self.src in attach since self.src gets value in _output_update when there's no metric. This makes problem in the case of calling attach again.

OK, let's just define `self.src = None` in the constructor and than `self.src` will be a metric if input `src` is a metric. This will help us to avoid doing `hasattr(self, "src")`, we can just do:
```
isinstance(self.src, Metric)
```

> By the way, why detach has not been implemented for RunningAverage?

Sorry, I do not remember

Comment (sadra-barikbin) [reply ignite/metrics/running_average.py:186]: Also why we don't apply `if not engine.has_event_handler` on `engine.add_event_handler(usage.COMPLETED, self.completed, name)` in `metric.py::attach`?

File: ignite/metrics/running_average.py [hunk 5]
```diff
@@ -18,8 +28,10 @@ class RunningAverage(Metric):
         alpha: running average decay factor, default 0.98
         output_transform: a function to use to transform the output if `src` is None and
             corresponds the output of process function. Otherwise it should be None.
-        epoch_bound: whether the running average should be reset after each epoch (defaults
-            to True).
+        epoch_bound: whether the running average should be reset after each epoch. It is depracated in favor of
+            ``usage`` argument in :meth:`attach` method. Setting ``epoch_bound`` to ``False`` is equivalent with
+            ``usage=SingleEpochRunningBatchWise()`` and setting it to ``True`` is equivalent with
```
Comment (vfdev-5) [inline ignite/metrics/running_average.py:33]: nit:
```suggestion
            ``usage`` argument in :meth:`attach` method. Setting ``epoch_bound`` to ``False`` is equivalent to
            ``usage=SingleEpochRunningBatchWise()`` and setting it to ``True`` is equivalent to
```

**Review by vfdev-5 — COMMENTED** [review]
@sadra-barikbin thanks for the update, there is a nit improvement for docstring. I'll check tests in more details a bit later.
Please also fix current problem with docs rendering:
https://app.netlify.com/sites/pytorch-ignite-preview/deploys/647cf10085384f0008aea77c
```
11:02:30 PM: reading sources... [100%] metrics .. utils
11:02:30 PM: /opt/build/repo/ignite/metrics/running_average.py:docstring of ignite.metrics.running_average.RunningAverage:8: ERROR: Unknown interpreted text role "method".
11:02:30 PM: looking for now-outdated files... none found
```

**Comment by sadra-barikbin** [comment]
> @sadra-barikbin thanks for the update, there is a nit improvement for docstring. I'll check tests in more details a bit later. Please also fix current problem with docs rendering: https://app.netlify.com/sites/pytorch-ignite-preview/deploys/647cf10085384f0008aea77c
> 
> ```
> 11:02:30 PM: reading sources... [100%] metrics .. utils
> 11:02:30 PM: /opt/build/repo/ignite/metrics/running_average.py:docstring of ignite.metrics.running_average.RunningAverage:8: ERROR: Unknown interpreted text role "method".
> 11:02:30 PM: looking for now-outdated files... none found
> ```

What do you mean by "nit"?

File: ignite/metrics/running_average.py (hunk 2 above)
Comment (sadra-barikbin) [reply ignite/metrics/running_average.py]: Could you please answer this?

**Comment by vfdev-5** [comment]
> > @sadra-barikbin thanks for the update, there is a nit improvement for docstring. I'll check tests in more details a bit later. Please also fix current problem with docs rendering: https://app.netlify.com/sites/pytorch-ignite-preview/deploys/647cf10085384f0008aea77c
> > ```
> > 11:02:30 PM: reading sources... [100%] metrics .. utils
> > 11:02:30 PM: /opt/build/repo/ignite/metrics/running_average.py:docstring of ignite.metrics.running_average.RunningAverage:8: ERROR: Unknown interpreted text role "method".
> > 11:02:30 PM: looking for now-outdated files... none found
> > ```
> 
> What do you mean by "nit"?

It's an expression to say a tiny improvement, I'm talking about this:https://github.com/pytorch/ignite/pull/2958#discussion_r1217809902

File: ignite/metrics/metric.py [hunk 6]
```diff
@@ -343,13 +434,17 @@ def completed(self, engine: Engine, name: str) -> None:
...
-                raise ValueError(f"usage should be 'EpochWise.usage_name' or 'BatchWise.usage_name', get {usage}")
+            for _usage in usages:
+                if usage == _usage.usage_name:
+                    usage = _usage()
+                    break
+            if isinstance(usage, str):
+                raise ValueError(
+                    "usage should be '(Running)EpochWise.usage_name' or "
+                    f"'((SingleEpoch)Running)BatchWise.usage_name', got {usage}"
+                )
```
Comment (vfdev-5) [inline ignite/metrics/metric.py:447]:
```
if isinstance(usage, str):
    usages = [EpochWise, RunningEpochWise, BatchWise, RunningBatchWise, SingleEpochRunningBatchWise]
    for usage_cls in usages:
        if usage == usage_cls.usage_name:
            usage = usage_cls()
            break
    if not isinstance(usage, MetricUsage):
```

File: ignite/metrics/running_average.py [hunk 7]
```diff
@@ -18,8 +28,10 @@ class RunningAverage(Metric):
         alpha: running average decay factor, default 0.98
         output_transform: a function to use to transform the output if `src` is None and
             corresponds the output of process function. Otherwise it should be None.
-        epoch_bound: whether the running average should be reset after each epoch (defaults
-            to True).
+        epoch_bound: whether the running average should be reset after each epoch. It is depracated in favor of
+            ``usage`` argument in :meth:`attach` method. Setting ``epoch_bound`` to ``False`` is equivalent to
+            ``usage=SingleEpochRunningBatchWise()`` and setting it to ``True`` is equivalent to
+            `usage=RunningBatchWise()` in the :meth:`attach` method. Default None.
```
Comment (vfdev-5) [inline ignite/metrics/running_average.py:34]:
```suggestion
            ``usage=RunningBatchWise()`` in the :meth:`attach` method. Default None.
```

File: ignite/metrics/running_average.py [hunk 8]
```diff
@@ -139,32 +151,101 @@ def compute(self) -> Union[torch.Tensor, float]:
...
+                                                                        batch is retrieved then.
+                ======================================================= ===========================================
+
+        ``RunningAverage`` retrieves ``engine.state.output`` at ``usage.ITERATION_COMPLETED`` if the ``src`` is not
+        given and it's computed and updated using ``src``, by manually calling its ``compute`` method, or
+        ``engine.state.output`` at ``usage.COMPLETED`` event.
+        Also if ``src`` is given, it is updated at ``usage.ITERATION_COMPLETED``, but its reset event is determined by
+        ``usage`` type. If ``isinstance(usage, BatchWise)`` holds true, ``src`` is reset on ``BatchWise().STARTED``,
+        otherwise on ``EpochWise().STARTED`` if ``isinstance(usage, EpochWise)``.
+        """
```
Comment (vfdev-5) [inline ignite/metrics/running_average.py:193]: Let's add 
```
.. versionchanged:: 0.5.1
    Added `usage` argument
```

File: ignite/metrics/running_average.py [hunk 9]
```diff
@@ -139,32 +151,101 @@ def compute(self) -> Union[torch.Tensor, float]:
...
+                "`epoch_bound` is deprecated and will be removed in the future. Consider using `usage` argument"
+                "instead. `epoch_bound=True` is equivalent with `usage=SingleEpochRunningBatchWise()` and "
+                "`epoch_bound=False` is equivalent with `usage=RunningBatchWise()`."
+            )
+            usage = SingleEpochRunningBatchWise() if self.epoch_bound else RunningBatchWise()
+
+        src_usage = EpochWise() if isinstance(usage, EpochWise) else BatchWise()
+        if isinstance(self.src, Metric) and not engine.has_event_handler(self.src.started, src_usage.STARTED):
+            engine.add_event_handler(src_usage.STARTED, self.src.started)
+            engine.add_event_handler(usage.ITERATION_COMPLETED, self.src.iteration_completed)
```
Comment (vfdev-5) [inline ignite/metrics/running_average.py:207]: @sadra-barikbin I'm not very sure to understand what you are doing here. Why do you need to define `src_usage`, can't 100% use `usage.STARTED` ? This looks a bit weird that we have to split like that `src_usage.STARTED` and `usage.ITERATION_COMPLETED`

File: tests/ignite/metrics/test_metric.py [hunk 10]
```diff
@@ -879,25 +893,52 @@ def _():
...
+        def _():
+            assert engine.state.metrics["rewm"] == 3 * sum(range(engine.state.epoch + 1))
+
+        engine.run([0, 1, 2], max_epochs=10)
+
+        m.detach(engine, usage=usage)
+
+    test("running_epoch_wise")
+    test(RunningEpochWise.usage_name)
+    test(RunningEpochWise())
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_metric.py:934]: Let's pytest parametrize here

File: ignite/metrics/running_average.py (hunk 9 above)
Comment (sadra-barikbin) [reply ignite/metrics/running_average.py:207]: We could evade this only if reinstate `epoch_bound` and hardcode start event of the very `RunningAverage` to be `Events.STARTED if not epoch_bound else Events.EPOCH_STARTED`. Then we could designate `usage.STARTED` for `src` start event. If we do this we no longer need `Running......` usage classes and the `EpochWise` and `BatchWise` themselves serve as usage.

File: ignite/metrics/running_average.py (hunk 2 above)
Comment (vfdev-5) [reply ignite/metrics/running_average.py]: > Also why we don't apply `if not engine.has_event_handler` on `engine.add_event_handler(usage.COMPLETED, self.completed, name)` in `metric.py::attach`?

Maybe, it is related to https://github.com/pytorch/ignite/pull/326#issuecomment-440463364

File: ignite/metrics/running_average.py (hunk 9 above)
Comment (vfdev-5) [reply ignite/metrics/running_average.py:210]: All this logic seems a bit complicated to me. I have to dive deeper into the code to understand all the details.

How about this ?
```suggestion
        if isinstance(self.src, Metric):
            src_usage = EpochWise() if isinstance(usage, EpochWise) else BatchWise()
            if not engine.has_event_handler(self.src.started, src_usage.STARTED):
                engine.add_event_handler(src_usage.STARTED, self.src.started)
            engine.add_event_handler(usage.ITERATION_COMPLETED, self.src.iteration_completed)
```

File: ignite/metrics/running_average.py (hunk 2 above)
Comment (sadra-barikbin) [reply ignite/metrics/running_average.py]: But the handlers in a metric don't accept args and kwargs. it's just the handler itself.

File: ignite/metrics/running_average.py (hunk 9 above)
Comment (sadra-barikbin) [reply ignite/metrics/running_average.py]: I changed the code. I think it is better now.

File: ignite/metrics/running_average.py [hunk 11]
```diff
@@ -101,70 +104,119 @@ def __init__(
...
-    def update(self, output: Sequence) -> None:
-        # Implement abstract method
-        pass
-
-    def compute(self) -> Union[torch.Tensor, float]:
-        if self._value is None:
-            self._value = self._get_src_value()
+    def update(self, output: Union[torch.Tensor, float]) -> None:
+        if self.src is None:
+            output = output.detach().to(self._device) if isinstance(output, torch.Tensor) else output
```
Comment (vfdev-5) [inline ignite/metrics/running_average.py:138]: Originally, the code here was:
```
            output = output.detach().to(self._device, copy=True)
```

File: ignite/metrics/running_average.py [hunk 12]
```diff
@@ -101,70 +104,119 @@ def __init__(
...
+            value = idist.all_reduce(output) / idist.get_world_size()
         else:
-            self._value = self._value * self.alpha + (1.0 - self.alpha) * self._get_src_value()
+            value = self.src.compute()
+            self.src.reset()
```
Comment (vfdev-5) [inline ignite/metrics/running_average.py:142]: I think this `self.src.reset()` call is redundant.

**Review by vfdev-5 — COMMENTED** [review]
Great work, thanks @sadra-barikbin 
I left few other comments to update the code and it will be good to go

File: ignite/metrics/running_average.py (hunk 11 above)
Comment (sadra-barikbin) [reply ignite/metrics/running_average.py:138]: I add `copy=True`. I think it has been added because the `idist.all_reduce` in the next line currently modifies its argument in-place if `output.device` be the same as that of the process.

File: ignite/metrics/running_average.py (hunk 12 above)
Comment (sadra-barikbin) [reply ignite/metrics/running_average.py:142]: But if we don't reset it, `src` keeps being updated till the end.

**Review by vfdev-5 — APPROVED** [review]
LGTM, thanks @sadra-barikbin !
